core/include/scipp/core/element_array.h

```diff
@@ -49,13 +49,36 @@ template <class T> class element_array {
     resize(new_size, default_init_elements);
   }
 
-  template <class InputIt,
-            std::enable_if_t<!std::is_integral<InputIt>{}, int> = 0>
-  element_array(InputIt first, InputIt last) {
+  //  template <class InputIt,
+  //            std::enable_if_t<!std::is_integral<InputIt>{}, int> = 0>
+  //  element_array(InputIt first, InputIt last) {
+  //    resize(std::distance(first, last), default_init_elements);
+  //    std::copy(first, last, data());
+  //  }
+
+  template <class Iter,
+            std::enable_if_t<
+                std::is_assignable<T &, typename Iter::value_type>{}, int> = 0>
+  element_array(Iter first, Iter last) {
     resize(std::distance(first, last), default_init_elements);
     std::copy(first, last, data());
   }
 
+  template <class U, std::enable_if_t<std::is_assignable_v<T &, U>, int> = 0>
+  element_array(U *first, U *last) {
+    resize(std::distance(first, last), default_init_elements);
+    std::copy(first, last, data());
+  }
```

Comment:
Can you avoid the duplication if you use something like `is_assignable_v<T &, decltype(*std::declval<Iter>())>`?